AEPCampaignClassic/Sources/CampaignClassicRegistrationManager.swift

```diff
@@ -0,0 +1,144 @@
+            return
+        }
+
+        /// bail out if the privacy is opted out.
+        let configuration = CampaignClassicConfiguration.init(forEvent: event, runtime: runtime)
+        guard configuration.privacyStatus == PrivacyStatus.optedIn else {
+            Log.debug(label: CampaignClassicConstants.LOG_TAG, "Device Registration failed, MobilePrivacyStatus is not optedIn.")
+            return
+        }
+
+        /// bail out if the required configuration for device registration request is unavailable
+        guard let integrationKey = configuration.integrationKey, let marketingServer = configuration.marketingServer else {
+            Log.debug(label: CampaignClassicConstants.LOG_TAG, "Device Registration failed, `campaignclassic.ios.integrationKey` and/or campaignclassic.marketingServer` configuration keys are unavailable.")
+            return
+        }
+
+        /// retrieve the userKey from the event
+        /// userKey is a string containing user identifier e.g. email
+        let userKey = event.userKey ?? ""
+        let additionalParametersXML = event.additionalParameters.serializeToXMLString()
+        let hashedData = String(format: "%@%@%@", deviceToken, userKey, additionalParametersXML).sha256()
+
+        /// bail out, If the registration request data has not changed
+        guard registrationInfoChanged(hashedData) else {
+            Log.debug(label: CampaignClassicConstants.LOG_TAG, "Device Registration dropped, there is no change in registration info since last successful request.")
+            return
+        }
+
+        /// build the payload
+        var payload = String(format: CampaignClassicConstants.REGISTRATION_PAYLOAD_FORMAT,
+                             URLEncoder.encode(value: deviceToken),
+                             URLEncoder.encode(value: integrationKey),
+                             URLEncoder.encode(value: userKey),
+                             URLEncoder.encode(value: UIDevice.current.name),
+                             URLEncoder.encode(value: UIDevice.current.model),
+                             URLEncoder.encode(value: CampaignClassicConstants.REGISTER_PARAM_DEVICE_BRAND_APPLE),
+                             URLEncoder.encode(value: CampaignClassicConstants.REGISTER_PARAM_DEVICE_MANUFACTURER_APPLE),
+                             URLEncoder.encode(value: UIDevice.current.systemName),
+                             URLEncoder.encode(value: systemInfoService.getFormattedOperatingSystem()),
+                             URLEncoder.encode(value: systemInfoService.getFormattedLocale()))
+        payload.append(additionalParametersXML)
+
+        /// build url and header
+        let urlString = String(format: CampaignClassicConstants.REGISTRATION_API_URL_BASE, marketingServer)
+        let headers = buildHeaders(payload: payload)
+        guard let url = URL(string: urlString) else {
```

Comment:
nit: move the guard statement above creation of `headers`, since we can't use `headers` without a valid `url`